### PR TITLE
expression: set collation id to negative in tipb if new collations are enabled

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve? 
This is a trick for the new collation implementations:

 - When new collations are enabled, we turn the collation id to negative so that other the
components of the cluster(for example, TiKV) is able to aware of it without any change to
the protocol definition.
 - When new collations are not enabled, collation id remains the same: all of them should be treated as `_bin` collations without padding.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has persistent data change

Side effects

 - Increased code complexity

Related changes

 - TiKV need to be updated.

Release note

 - NA

### CODE REVIEW

File: util/collate/collate.go [hunk 1]
```diff
@@ -90,6 +90,26 @@ func NewCollationEnabled() bool {
 	return newCollationEnabled
 }
 
+// RewriteNewCollationIDIfNeeded rewrites a collation id if the new collations are enabled.
+// When new collations are enabled, we turn the collation id to negative so that other the
+// components of the cluster(for example, TiKV) is able to aware of it without any change to
+// the protocol definition.
+// When new collations are not enabled, collation id remains the same.
+func RewriteNewCollationIDIfNeeded(id int32) int32 {
+	if newCollationEnabled && id > 0 {
```
Comment (wjhuang2016) [inline util/collate/collate.go]: `id > 0` can be removed?

File: expression/expr_to_pb_test.go [hunk 2]
```diff
@@ -673,3 +674,50 @@ func (s *testEvaluatorSerialSuites) TestMetadata(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(metadata.InUnion, Equals, true)
 }
+
+func columnCollation(c *Column, coll string, flen int) *Column {
+	c.RetType.Collate = coll
+	c.RetType.Flen = flen
```
Comment (wjhuang2016) [inline expression/expr_to_pb_test.go]: We can ignore `Flen` here.

File: util/collate/collate.go (hunk 1 above)
Comment (bb7133) [reply util/collate/collate.go]: This is a guarantee for unexpected situation. I think we can keep it.

File: expression/expr_to_pb_test.go (hunk 2 above)
Comment (bb7133) [reply expression/expr_to_pb_test.go]: Done.

File: util/collate/collate.go (hunk 1 above)
Comment (bb7133) [reply util/collate/collate.go]: I've added a warning here.

**Pull request merged**